File: backend-go/internal/models/invoice.go
// package models to accomodate models
package models

import (
	"time"

	"github.com/google/uuid"
)

type RawInvoice struct {
	ID        uuid.UUID `json:"id"`
	ImageData []byte    `json:"image_data"`
	CreatedAt time.Time `json:"created_at"`
}

type ParsedInvoice struct {
	ID            uuid.UUID    `json:"id"`
	RawID         uuid.UUID    `json:"raw_id"`
	RawText       *string      `json:"raw_text"`
	AccessKey     *string      `json:"access_key"`
	Establishment *string      `json:"establishment"`
	CNPJ          *string      `json:"cnpj"`
	Date          *time.Time   `json:"date"`
	Total         *float64     `json:"total"`
	Items         []ParsedItem `json:"items"`
	ParserVersion string       `json:"parser_version"`
	IsValid       bool         `json:"is_valid"`
	ParseErrors   []string     `json:"parse_errors"`
}

type ParsedItem struct {
	Name      *string  `json:"name"`
	Quantity  *int     `json:"quantity"`
	UnitPrice *float64 `json:"unit_price"`
	Total     *float64 `json:"total"`
}

func (p *ParsedInvoice) Validate() {
	p.IsValid = true
	p.ParseErrors = nil

	if p.CNPJ == nil {
		p.IsValid = false
		p.ParseErrors = append(p.ParseErrors, "CNPJ not identified")
	}

	if len(p.Items) == 0 {
		p.IsValid = false
		p.ParseErrors = append(p.ParseErrors, "Invoice items not identified")
	}

	if p.Total == nil || *p.Total <= 0 {
		p.IsValid = false
		p.ParseErrors = append(p.ParseErrors, "Total price not identified")
	}

	if p.Date == nil {
		p.IsValid = false
		p.ParseErrors = append(p.ParseErrors, "Invoice date not identified")
	}
}

File: frontend-avalonia/Models/Invoice.cs
using System;
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace frontend_avalonia.Models;

public class Invoice
{
    [JsonPropertyName("id")]
    public Guid Id { get; set; }

    [JsonPropertyName("raw_id")]
    public Guid RawId { get; set; }

    [JsonPropertyName("raw_text")]
    public string? RawText { get; set; }

    [JsonPropertyName("access_key")]
    public string? AccessKey { get; set; }

    [JsonPropertyName("establishment")]
    public string? Establishment { get; set; }

    [JsonPropertyName("cnpj")]
    public string? Cnpj { get; set; }

    [JsonPropertyName("date")]
    public DateTime? Date { get; set; }

    [JsonPropertyName("total")]
    public decimal? Total { get; set; }

    [JsonPropertyName("items")]
    public List<ParsedItem> Items { get; set; } = new();

    [JsonPropertyName("parser_version")]
    public string ParserVersion { get; set; } = string.Empty;

    [JsonPropertyName("is_valid")]
    public bool IsValid { get; set; }

    [JsonPropertyName("parse_errors")]
    public List<string> ParseErrors { get; set; } = new();
}